File: src/main/java/org/pl/parser/exception/ParserException.java
package org.pl.parser.exception;

import org.pl.parser.ParserError;

import java.util.List;

/**
 * A parser throws this exception if any errors occurred during token analysis.
 */
public class ParserException extends RuntimeException {

    private List<ParserError> errors;

    public ParserException(String message, List<ParserError> errors) {
        super(message);
        this.errors = errors;
    }

    public List<ParserError> getErrors() {
        return errors;
    }

    @Override
    public String toString() {
        var result = new StringBuilder(128);
        result.append(ParserException.class);
        result.append(": ");
        result.append(getMessage());
        result.append("\n");
        for (ParserError error : errors) {
            result.append(" - ");
            result.append(error);
        }
        result.append("\n");
        return result.toString();
    }
}

File: src/main/java/org/pl/parser/exception/ParserException.kt
package org.pl.parser.exception

import org.pl.parser.ParserError

/**
 * A parser throws this exception if any errors occurred during token analysis.
 */
class ParserException(message: String?, private val errors: List<ParserError>) : RuntimeException(message) {
    override fun toString() = "The parser detected some semantic errors and therefore failed (ParserException): ${
        errors
                .map { "\n$it" }
                .reduce { acc, s -> acc + s }
    }"
}